datafusion/optimizer/src/optimizer.rs

```diff
@@ -97,13 +106,85 @@ impl Optimizer {
         debug!("Input logical plan:\n{}\n", plan.display_indent());
         trace!("Full input logical plan:\n{:?}", plan);
         for rule in &self.rules {
-            new_plan = rule.optimize(&new_plan, optimizer_config)?;
-            observer(&new_plan, rule.as_ref());
-            debug!("After apply {} rule:\n", rule.name());
-            debug!("Optimized logical plan:\n{}\n", new_plan.display_indent());
+            let result = rule.optimize(&new_plan, optimizer_config);
+            match result {
+                Ok(plan) => {
+                    new_plan = plan;
+                    observer(&new_plan, rule.as_ref());
+                    debug!("After apply {} rule:\n", rule.name());
+                    debug!("Optimized logical plan:\n{}\n", new_plan.display_indent());
+                }
+                Err(ref e) => {
+                    if optimizer_config.skip_failing_rules {
```

Comment:
```suggestion
                    if optimizer_config.skip_failing_rules {
                      // Note to future readers: if you see this warning it signals a
                      // bug in the DataFusion optimizer. Please consider filing a ticket
                      // https://github.com/apache/arrow-datafusion
```

I think it might be good to tell anyone who sees this error that it is not expected and that it means there is a bug somewhere in datafusion